hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java

```diff
@@ -351,38 +260,62 @@ protected boolean checkIfContainerToExclude(Throwable t) {
         && t instanceof ContainerNotOpenException;
   }
 
-  void writeParityCells(int parityCellSize) throws IOException {
-    final ByteBuffer[] buffers = ecChunkBufferCache.getDataBuffers();
+  private void generateParityCells() throws IOException {
+    final ByteBuffer[] dataBuffers = ecChunkBufferCache.getDataBuffers();
     final ByteBuffer[] parityBuffers = ecChunkBufferCache.getParityBuffers();
 
+    // parityCellSize = min(ecChunkSize, stripeSize)
+    //                = min(cellSize, sum(dataBuffer positions))
+    //                = min(dataBuffer[0].limit(), dataBuffer[0].position())
+    //                = dataBuffer[0].position()
+    final int parityCellSize = dataBuffers[0].position();
+
+    // Duplicate dataBuffers and add padding
```

Comment:
Isn't this buffers we are creating out side of cache?

I think this may be a performance impact if there are many small files?
Pretty much for every file we will endup creating this buffer once outside of cache.

Reply:
The `duplicate` don't create extra buffer, it shares same underlying buffer with the original ByteBuffer.

Reply:
Or can we add `dataBuffersForRead` in `ECChunkBuffers` to avoid the object creation cost here?

Reply:
Yeah you are right. Also this will happen sequentially so underlying buffer sharing is ok.
IIUC, How about we do this only when padding is necessary? Otherwise we may can use original buffers when all databuffers actually full right? So, min size of databuffers should tell use whether we will need padding or not right?
That way this extra object will be created only for last stripe right.

Reply:
I don't think it's safe to use the original buffers here because `encoder.encode()` will modify the position of the input ByteBuffer.

Reply:
Do you think adding a `dataBufferForEncode` like https://github.com/kaijchen/ozone/commit/b15799b447708033740dd9f105954217b86d5e31 is OK?

Reply:
That would too much usage of cache buffers class than needed.
>IIUC, How about we do this only when padding is necessary? Otherwise we may can use original buffers when all databuffers actually full right? 

Not sure you got what I was trying to saying. Why original buffers will be modified when original buffers already full (all full with ecChunkSize)?
I was suggesting to create this duplicate buffers only when we really do padding. Let's say if all data buffers are full, then we will not do any padding. But still we create this duplicate right.

>I don't think it's safe to use the original buffers here because encoder.encode() will modify the position of the input ByteBuffer.
Just for argument: Ideally you will modify original buffer here also. But we will have our own tracking pointers. Similar way, when you change the original buffer directly and we track that length, we are ok too.

Reply:
OK, I got your point.

Reply:
I have added an extra stripe state check in the new commit.
And instead of duplicating the dataBuffer, we can restore the positions if it's in legal state.